File: mars-rover/python/tests/rover_builder.py
from mars_rover import (
    DEFAULT_GRID_HEIGHT,
    DEFAULT_GRID_WIDTH,
    Direction,
    MovementStatus,
    Rover,
)


class RoverBuilder:
    def __init__(self) -> None:
        self._x = 0
        self._y = 0
        self._heading = Direction.NORTH
        self._grid_width = DEFAULT_GRID_WIDTH
        self._grid_height = DEFAULT_GRID_HEIGHT
        self._obstacles: set[tuple[int, int]] = set()

    def at(self, x: int, y: int) -> "RoverBuilder":
        self._x = x
        self._y = y
        return self

    def facing(self, heading: Direction) -> "RoverBuilder":
        self._heading = heading
        return self

    def on_grid(self, width: int, height: int) -> "RoverBuilder":
        self._grid_width = width
        self._grid_height = height
        return self

    def with_obstacle_at(self, x: int, y: int) -> "RoverBuilder":
        self._obstacles.add((x, y))
        return self

    def build(self) -> Rover:
        return Rover(
            self._x, self._y, self._heading,
            self._grid_width, self._grid_height,
            set(self._obstacles),
            MovementStatus.MOVING, None,
        )

File: mars-rover/typescript/tests/roverBuilder.ts
import {
  DEFAULT_GRID_WIDTH,
  DEFAULT_GRID_HEIGHT,
  type Direction,
  Rover,
} from '../src/rover.js';

export class RoverBuilder {
  private x = 0;
  private y = 0;
  private heading: Direction = 'North';
  private gridWidth = DEFAULT_GRID_WIDTH;
  private gridHeight = DEFAULT_GRID_HEIGHT;
  private readonly obstacles = new Set<string>();

  at(x: number, y: number): this { this.x = x; this.y = y; return this; }
  facing(heading: Direction): this { this.heading = heading; return this; }
  onGrid(width: number, height: number): this { this.gridWidth = width; this.gridHeight = height; return this; }
  withObstacleAt(x: number, y: number): this { this.obstacles.add(`${x},${y}`); return this; }

  build(): Rover {
    return new Rover({
      x: this.x, y: this.y, heading: this.heading,
      gridWidth: this.gridWidth, gridHeight: this.gridHeight,
      obstacles: new Set(this.obstacles),
      status: 'Moving', lastObstacle: null,
    });
  }
}
